// devexpress-reporting/designer/tools/smartTags/expressionSmartTag.js
/**
* DevExpress HTML/JS Reporting (designer\tools\smartTags\expressionSmartTag.js)
* Version:  22.2.14
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { Disposable } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { ReportExpressionEditorWrapper } from '../../widgets/expressioneditor/reportExpressionEditorWrapper';
export class ExpressionSmartTag extends Disposable {
    constructor(reportElement, _expressionEditor) {
        super();
        this.reportElement = reportElement;
        this._expressionEditor = _expressionEditor;
        this.imageTemplateName = 'dxrd-svg-properties-propertyexpression';
    }
    onClick() {
        var expressionEditor = new ReportExpressionEditorWrapper(ko.observable(this.reportElement), ko.observable(null));
        this._expressionEditor(expressionEditor);
        this._expressionEditor().popupVisible(true);
        this.subscription = expressionEditor.popupVisible.subscribe((newVal) => {
            if (!newVal) {
                setTimeout(() => {
                    expressionEditor.dispose();
                    this._expressionEditor(null);
                    this.subscription.dispose();
                }, 10);
            }
        });
    }
    dispose() {
        var _a, _b;
        (_a = this._expressionEditor()) === null || _a === void 0 ? void 0 : _a.dispose();
        this._expressionEditor(null);
        (_b = this.subscription) === null || _b === void 0 ? void 0 : _b.dispose();
    }
}
