// @devexpress/analytics-core/widgets/ace/_binding.js
/**
* DevExpress Analytics (widgets\ace\_binding.js)
* Version:  22.2.14
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import ace from 'ace-builds/src-noconflict/ace';
import languageTools from 'ace-builds/src-noconflict/ext-language_tools';
import modeJson from 'ace-builds/src-noconflict/mode-json';
import modeSql from 'ace-builds/src-noconflict/mode-sql';
import modeText from 'ace-builds/src-noconflict/mode-text';
import themeAmbiance from 'ace-builds/src-noconflict/theme-ambiance';
import themeDreamweaver from 'ace-builds/src-noconflict/theme-dreamweaver';
import * as $ from 'jquery';
import * as ko from 'knockout';
import { addDisposeCallback } from '../../serializer/_internal';
import { isDarkTheme } from '../_utils';
import { aceAvailable } from './_ace-available';
import { defineAceCriteria } from './_ace-mode-criteria';
import { defineAceDocComments } from './_ace-mode-doc-comment';
var aceDefined = false;
export function defineAce(ace) {
    if (!aceDefined) {
        defineAceDocComments(ace);
        defineAceCriteria(ace);
        if ('setModuleLoader' in ace.config) {
            ace.config.setModuleLoader('./theme/textmate', () => __awaiter(this, void 0, void 0, function* () { return ace.require('./theme/textmate'); }));
            ace.config.setModuleLoader('ace/mode/criteria', () => __awaiter(this, void 0, void 0, function* () { return ace.require('ace/mode/criteria'); }));
            ace.config.setModuleLoader('ace/theme/dreamweaver', () => __awaiter(this, void 0, void 0, function* () { return ace.require('ace/theme/dreamweaver'); }));
            ace.config.setModuleLoader('ace/theme/ambiance', () => __awaiter(this, void 0, void 0, function* () { return ace.require('ace/theme/ambiance'); }));
            ace.config.setModuleLoader('ace/mode/sql', () => __awaiter(this, void 0, void 0, function* () { return ace.require('ace/mode/sql'); }));
            ace.config.setModuleLoader('ace/mode/text', () => __awaiter(this, void 0, void 0, function* () { return ace.require('ace/mode/text'); }));
            ace.config.setModuleLoader('ace/mode/json', () => __awaiter(this, void 0, void 0, function* () { return ace.require('ace/mode/json'); }));
        }
        if (!window['ace'] && aceAvailable()) {
            ace.config.setModuleUrl('ace/ext/language_tools', languageTools);
            ace.config.setModuleUrl('ace/mode/sql', modeSql);
            ace.config.setModuleUrl('ace/mode/text', modeText);
            ace.config.setModuleUrl('ace/mode/json', modeJson);
            ace.config.setModuleUrl('ace/theme/ambiance', themeAmbiance);
            ace.config.setModuleUrl('ace/theme/dreamweaver', themeDreamweaver);
        }
    }
    aceDefined = true;
}
ko.bindingHandlers['dxAceEditor'] = {
    init: function (element, valueAccessor, allBindings, viewModel, bindingContext) {
        var _ace = ace || window['ace'];
        defineAce(_ace);
        var values = valueAccessor(), text = values.value, editorContainer = values.editorContainer, editor, shouldProcessOnChangeEvent = true, _setEditorText = (editorInstance, text) => {
            shouldProcessOnChangeEvent = false;
            editorInstance.getSession().setValue((text && text.toString()) || '');
            editorInstance.clearSelection();
            editorInstance.getSession().getUndoManager().reset();
            shouldProcessOnChangeEvent = true;
        };
        if (_ace) {
            var showGutter = values.options.showGutter != undefined ? values.options.showGutter : true;
            var additionalOptions = values.additionalOptions;
            var langTools = _ace.require('ace/ext/language_tools');
            editor = _ace.edit(element);
            var guid = ko.observable(null);
            var theme = values.theme;
            if (!theme)
                theme = isDarkTheme() ? 'ace/theme/ambiance' : 'ace/theme/dreamweaver';
            editor.setTheme(theme);
            editor.$blockScrolling = Infinity;
            var languageMode = viewModel.languageHelper.getLanguageMode();
            var session = editor.getSession();
            session.gutterRenderer = {
                getWidth: (session, lastLineNumber, config) => {
                    return lastLineNumber.toString().length * config.characterWidth;
                },
                getText: (session, row) => {
                    return row + 1;
                }
            };
            session.setMode(languageMode);
            var onBlur = () => { editor.completer && editor.completer.popup && editor.completer.popup.hide(); }, onChange, onFocus, onChangeAnnotation;
            if (additionalOptions) {
                if (additionalOptions.onChange) {
                    var timer = null;
                    onChange = (e) => {
                        if (timer !== null)
                            clearTimeout(timer);
                        if (shouldProcessOnChangeEvent) {
                            timer = setTimeout(() => {
                                if (text() !== session.getValue() || (session.getAnnotations() || []).some(x => x.type === 'error')) {
                                    additionalOptions.onChange(session);
                                }
                            }, additionalOptions && additionalOptions.changeTimeout || 1000);
                        }
                    };
                    session.on('change', onChange);
                }
                if (additionalOptions.overrideEditorFocus) {
                    editor.focus = function (a, e) {
                        editor.textInput.getElement().focus();
                    };
                }
                if (additionalOptions.onFocus) {
                    onFocus = () => { additionalOptions.onFocus(session); };
                    editor.on('focus', onFocus);
                }
                if (additionalOptions.onBlur) {
                    onBlur = () => {
                        editor.completer && editor.completer.popup && editor.completer.popup.hide();
                        return additionalOptions.onBlur(session);
                    };
                }
                if (additionalOptions.onChangeAnnotation) {
                    onChangeAnnotation = (a, e) => {
                        additionalOptions.onChangeAnnotation(e);
                    };
                    session.on('changeAnnotation', onChangeAnnotation);
                }
            }
            editor.on('blur', onBlur);
            var completers = viewModel.languageHelper.createCompleters(editor, bindingContext, viewModel);
            langTools.setCompleters(completers);
            editor.setOptions(Object.assign({ autoScrollEditorIntoView: false }, values.options));
            if (additionalOptions && 'setUseWrapMode' in additionalOptions) {
                editor.getSession().setUseWrapMode(additionalOptions.setUseWrapMode);
            }
            if (!showGutter) {
                editor.renderer.setShowGutter(showGutter);
            }
            if (editor.renderer.$gutter) {
                var gutterClassName = editor.renderer.$gutter.className + ' dxd-border-primary dxd-text-primary dxd-back-primary';
                editor.renderer.$gutter.className = gutterClassName;
            }
            var oldMouseMove = editor._defaultHandlers.guttermousemove;
            editor._defaultHandlers.guttermousemove = function (e) {
                var rect = element.getBoundingClientRect();
                e.x = e.x - rect.left;
                e.y = e.y - rect.top;
                oldMouseMove(e);
            };
            if (ko.isSubscribable(text)) {
                var subscription = text.subscribe((newText) => {
                    if (newText !== session.getValue()) {
                        _setEditorText(editor, newText);
                    }
                    if (additionalOptions && additionalOptions.onValueChange)
                        additionalOptions.onValueChange(editor);
                });
            }
            _setEditorText(editor, ko.unwrap(text));
            if (values.callbacks)
                values.callbacks.focus = () => {
                    setTimeout(() => {
                        editor.textInput.getElement().focus();
                    }, 10);
                };
            addDisposeCallback(element, function () {
                editor.completers && editor.completers.splice(0);
                editor._defaultHandlers.guttermousemove = oldMouseMove;
                subscription.dispose();
                if (values.callbacks)
                    values.callbacks.focus = $.noop;
                completers.forEach(x => x.dispose && x.dispose());
                onBlur && editor.off('blur', onBlur);
                onFocus && editor.off('focus', onFocus);
                onChange && session.off('change', onChange);
                onChangeAnnotation && session.off('changeAnnotation', onChangeAnnotation);
                editor.destroy();
            });
        }
        if (ko.isObservable(editorContainer)) {
            editorContainer(editor);
        }
    }
};
