// devexpress-reporting/designer/actions/reportActions.js
/**
* DevExpress HTML/JS Reporting (designer\actions\reportActions.js)
* Version:  22.2.14
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { Disposable, getLocalization } from '@devexpress/analytics-core/analytics-utils';
import * as ko from 'knockout';
import { defaultCulture } from '../../common/metadata';
import { BandViewModel } from '../bands/xrBand';
import { ReportViewModel } from '../controls/xrReport';
export class ReportActions extends Disposable {
    constructor(onComponentAdded) {
        super();
        this._contextModel = ko.observable();
        this._disposables.push(this._targetModel = ko.computed(() => {
            var model = this._contextModel();
            if (model) {
                if (model.parentModel() && model.parentModel().parentModel() && model.controlType === 'SubBand') {
                    model = model.parentModel().parentModel();
                }
                else if (model.parentModel() && model.controlType !== 'DevExpress.XtraReports.UI.XtraReport' && model.controlType !== 'DetailReportBand') {
                    model = model.parentModel();
                }
            }
            return model;
        }));
        this.actions = [
            {
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                text: 'Insert Top Margin Band',
                displayText: () => getLocalization('Insert Top Margin Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertTopMarginBand'),
                imageClassName: 'dxrd-image-actions-top_margin',
                imageTemplateName: 'dxrd-svg-actions-top_margin',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('TopMarginBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('TopMarginBand'); },
            }, {
                text: 'Insert Report Header Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Report Header Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertReportHeaderBand'),
                imageClassName: 'dxrd-image-actions-report_header',
                imageTemplateName: 'dxrd-svg-actions-report_header',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('ReportHeaderBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('ReportHeaderBand'); },
            }, {
                text: 'Insert Page Header Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Page Header Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertPageHeaderBand'),
                imageClassName: 'dxrd-image-actions-page_header',
                imageTemplateName: 'dxrd-svg-actions-page_header',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('PageHeaderBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('PageHeaderBand'); },
            }, {
                text: 'Insert Group Header Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Group Header Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertGroupHeaderBand'),
                imageClassName: 'dxrd-image-actions-group_header',
                imageTemplateName: 'dxrd-svg-actions-group_header',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('GroupHeaderBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('GroupHeaderBand'); },
            }, {
                text: 'Insert Detail Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Detail Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertDetailBand'),
                imageClassName: 'dxrd-image-actions-detail',
                imageTemplateName: 'dxrd-svg-actions-detail',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('DetailBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('DetailBand'); },
            }, {
                text: 'Insert Vertical Header Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Vertical Header Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertVerticalHeaderBand'),
                imageClassName: 'dxrd-image-actions-vertical_header',
                imageTemplateName: 'dxrd-svg-actions-vertical_header',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('VerticalHeaderBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('VerticalHeaderBand'); },
            }, {
                text: 'Insert Vertical Detail Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Vertical Detail Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertVerticalDetailBand'),
                imageClassName: 'dxrd-image-actions-vertical_detail',
                imageTemplateName: 'dxrd-svg-actions-vertical_detail',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('VerticalDetailBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('VerticalDetailBand'); },
            }, {
                text: 'Insert Vertical Total Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Vertical Total Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertVerticalTotalBand'),
                imageClassName: 'dxrd-image-actions-vertical_total',
                imageTemplateName: 'dxrd-svg-actions-vertical_total',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('VerticalTotalBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('VerticalTotalBand'); },
            }, {
                text: 'Insert Detail Report Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Detail Report Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertDetailReportBand'),
                imageClassName: 'dxrd-image-actions-detail_report',
                imageTemplateName: 'dxrd-svg-actions-detail_report',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('DetailReportBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('DetailReportBand'); },
            }, {
                text: 'Insert Group Footer Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Group Footer Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertGroupFooterBand'),
                imageClassName: 'dxrd-image-actions-group_footer',
                imageTemplateName: 'dxrd-svg-actions-group_footer',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('GroupFooterBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('GroupFooterBand'); },
            }, {
                text: 'Insert Report Footer Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Report Footer Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertReportFooterBand'),
                imageClassName: 'dxrd-image-actions-report_footer',
                imageTemplateName: 'dxrd-svg-actions-report_footer',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('ReportFooterBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('ReportFooterBand'); },
            }, {
                text: 'Insert Page Footer Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Page Footer Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertPageFooterBand'),
                imageClassName: 'dxrd-image-actions-page_footer',
                imageTemplateName: 'dxrd-svg-actions-page_footer',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('PageFooterBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('PageFooterBand'); },
            }, {
                text: 'Insert Bottom Margin Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Bottom Margin Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertBottomMarginBand'),
                imageClassName: 'dxrd-image-actions-bottom_margin',
                imageTemplateName: 'dxrd-svg-actions-bottom_margin',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('BottomMarginBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('BottomMarginBand'); },
            }, {
                text: 'Insert Sub-Band',
                group: () => getLocalization('Bands', 'ReportStringId.RibbonXRDesign_PageGroup_Bands'),
                displayText: () => getLocalization('Insert Sub-Band', 'ASPxReportsStringId.ReportDesigner_ReportActions_InsertSubBand'),
                imageClassName: 'dxrd-image-actions-subband',
                imageTemplateName: 'dxrd-svg-actions-subband',
                disabled: ko.pureComputed(() => {
                    return !this._canAddBand('SubBand');
                }),
                visible: true,
                clickAction: () => { this._addBand('SubBand'); },
            }
        ];
        this.onComponentAdded = (e) => { onComponentAdded && onComponentAdded(e); };
    }
    _canAddBand(bandType) {
        if (!this._targetModel()) {
            return false;
        }
        var report = this._targetModel().root;
        if (report && report.language && report.language() !== defaultCulture)
            return false;
        if (bandType === 'GroupHeaderBand' || bandType === 'GroupFooterBand' || bandType === 'DetailReportBand') {
            return true;
        }
        if (bandType === 'SubBand') {
            var model = (this._contextModel());
            return model !== null && !model.lockedInUserDesigner() && (model.controlType.indexOf('XtraReport') === -1 && ['DetailReportBand', 'TopMarginBand', 'BottomMarginBand', 'SubBand'].indexOf(model.controlType) === -1);
        }
        if (this._targetModel().controlType === 'DetailReportBand' && (bandType === 'TopMarginBand' || bandType === 'BottomMarginBand' || bandType === 'PageHeaderBand' || bandType === 'PageFooterBand')) {
            return false;
        }
        return this._targetModel()['bands']().filter((band) => band.controlType === bandType).length === 0;
    }
    _addBand(bandType) {
        if (this._canAddBand(bandType)) {
            var model = this._targetModel();
            if (bandType === 'SubBand') {
                model = this._contextModel();
            }
            var height = model.root['dpi']();
            var control = model.createChild({ '@ControlType': bandType, '@HeightF': height });
            this.onComponentAdded({ parent: model, model: control });
        }
    }
    getActions(context) {
        if (context instanceof ReportViewModel || context instanceof BandViewModel) {
            this._contextModel(context);
            return this.actions;
        }
        else {
            this._contextModel(null);
        }
        return [];
    }
}
